### PR TITLE
Some fixes to mir-borrowck

### PR DESCRIPTION
**Pull request by pnkfelix**
Make the code more closely match the NLL RFC (updated description).

(The biggest visible fix the addition of the Shallow/Deep distinction, which means mir-borrowck stops falsely thinking that StorageDeads need deep access to their input L-value.)

### CODE REVIEW

**Comment by pnkfelix** [comment]
@bors r- 

I want to investigate the StorageDead change more after pushback from @arielb1

**Comment by pnkfelix** [comment]
(one further refactoring that I am considering is to fold the calls to `check_if_path_is_moved` *into* the body of `access_lvalue`. There is one case makes that a non-trivial refactoring, however, so I am leaving it for later work.)

File: src/librustc_mir/borrow_check.rs [hunk 1]
```diff
@@ -174,13 +174,17 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> DataflowResultsConsumer<'b, 'gcx>
         match stmt.kind {
             StatementKind::Assign(ref lhs, ref rhs) => {
                 self.mutate_lvalue(ContextKind::AssignLhs.new(location),
-                                   (lhs, span), JustWrite, flow_state);
+                                   (lhs, span), Shallow(None), JustWrite, flow_state);
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: AST borrowck basically does a `Deep` write here and we might want to stay in sync with it. Also, if you leave this as `Shallow` without erasing borrows in `dataflow::impls::borrows` you'll have some *very* confusing diagnostics.

File: src/librustc_mir/borrow_check.rs [hunk 2]
```diff
@@ -309,29 +320,116 @@ enum ConsumeKind { Drop, Consume }
...
+                        }
+                        Control::Break
+                    }
+                    (Write(kind), _) => {
+                        match kind {
+                            WriteKind::MutableBorrow(bk) =>
+                                this.report_conflicting_borrow(context, lvalue_span,
+                                                               (lvalue_span.0, bk),
+                                                               (&borrow.lvalue, borrow.kind)),
+                            WriteKind::StorageDead | // separate diagnostic for StorageDead case?
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: In AST borrowck, this is "borrow does not live long enough" for both `Drop` and `StorageDead`. ofc. you can match on the context instead if you want.

File: src/librustc_mir/borrow_check.rs [hunk 3]
```diff
@@ -309,29 +320,116 @@ enum ConsumeKind { Drop, Consume }
...
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+enum WriteKind {
+    StorageDead,
+    MutableBorrow(BorrowKind),
+    Mutate,
+    Move,
+}
+
 impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx> {
+    fn access_lvalue(&mut self,
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: You should check somewhere that you have sufficient permissions to do the access - aka no mutable borrows of immutable references or moves through a non-`Box` reference.

Comment (arielb1) [reply src/librustc_mir/borrow_check.rs]: You can leave that to a different PR through.

**Comment by arielb1** [comment]
Basically r=me after a rebase. There are a few comments, but you can leave them to me or to a follow-up PR.

File: src/librustc_mir/borrow_check.rs [hunk 4]
```diff
@@ -664,228 +694,179 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
...
-            // loans; i.e. it took a scope S and checked that each
-            // restriction's kill_scope was a superscope of S.
+        'next_borrow: for i in flow_state.borrows.elems_incoming() {
             let borrowed = &data[i];
-            for restricted in self.restrictions(&borrowed.lvalue) {
-                if restricted == lvalue {
+
+            // Is `lvalue` (or a prefix of it) already borrowed? If
+            // so, that's relevant.
+            for accessed_prefix in self.prefixes(lvalue, PrefixSet::All) {
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: This includes a "drive-by" fix to #38899. I think we might have a "legacy compatibility mode" that does not include this drive-by fix, so we can get a good summary of real issues in a crater.

At least add a comment `// DIFFERENCE FROM BORROWCK`.

**Comment by pnkfelix** [comment]
Thanks for taking time over your weekend for that review @arielb1 ! Rebasing now!

**Comment by pnkfelix** [comment]
I think my approach to the AST-/MIR-borrrowck discrepancies that @arielb1 has pointed out will be to just put in comments for now. Haven't decided 100% about that yet, but since one has to opt in to using MIR-borrowck in the first place via a `-Z` switch, it seems reasonable to delay porting the AST-borrowck "bugs", at least until we have things like two-phase borrows implemented.

File: src/librustc_mir/borrow_check.rs (hunk 3 above)
Comment (pnkfelix) [reply src/librustc_mir/borrow_check.rs]: (added a fixme-comment noting that we need to have this.)
